_posts/2023-11-01-#28368.md

```diff
@@ -7,19 +7,61 @@ authors: [ismaelsadeeq]
+
+* The [`CValidationInterface`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/validationinterface.h#L77) is a class any subscriber can implement to receive event callbacks whenever they are generated.
+   * The callbacks are generated from validation and mempool, `CValidationInterface` has callbacks for major events that are happening in the node, when a transaction is added to the mempool [`TransactionAddedToMempool`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/validationinterface.h#L99), when a transaction is removed from the mempool [`TransactionRemovedFromMempool`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/validationinterface.h#L133), when a new block is connected to the chain [`BlockConnected`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/validationinterface.h#L140) and others.
+
+   * The callbacks will be received by the subscriber in the order in which they are generated by validation and mempool.
+
+* This PR decouples the fee estimator from the mempool by implementing the `CValidationInterface` and subscribing to events relevant to the fee estimator.
+
+* The PR added another `CValidationInterface` callback`MempoolTransactionsRemovedForConnectedBlock` which is emitted in  [`removeForBlock`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/txmempool.h#L487).
+* The PR also modifies the [`TransactionAddedToMempool`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/validationinterface.h#L99) callback parameter from `CTransactionRef` to a new struct which has the [`CTransactionRef`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/primitives/transaction.h#L421) and other information the fee estimator needs such as the size, fees, and others. There has been an attempt at this PR [#11775](https://github.com/bitcoin/bitcoin/pull/11775)
 ## Questions
+
+### Conceptual
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+2. Why is it beneficial to decouple the fee estimator from the mempool?
+3.  Are there any benefits of the fee estimator coupled to the mempool? Are there downsides to decoupling it?
```

Comment:
Maybe add a few questions about the `CValidationInterface` as it could be something people struggle with. e.g.

- Can you list the other subscribers to the interface? (Hint: look for classes that inherit from `CValidationInterface`)
- What does it mean for a class to implement one of the `CValidationInterface` methods? Why is it similar to "subscribing to the event?"
- What is the difference between implementing a `BlockConnected` and a `NewPoWValidBlock` callback, apart from the fact that they are different events? On which thread is each callback executed?

Reply:
I added the questions, thanks for ideas.